### DISCUSSION TITLE
Install uasyncio on esp8266

### DISCUSSION THREAD

**Discussion by ShocKwav3** [micropython/micropython#12762] [ESP8266] [answered]
Hi!

Been trying to get this done but strangely enough I fail to see how. I tried with mip but it does not work probably because uasyncio is not available where mip looks. I can't use either upip because upip is not provided with the micropython build, maybe because mine is a 1mb esp8266.
I have mpy-cross ready so I can compile the uasyncio but I cant find what file to compile.
I found [this discussion](https://github.com/orgs/micropython/discussions/9350) where a comment to the answer suggests building a firmware myself with what I want could be an option. I did not find any instruction on how could I build my own firmware.

So, now I have two questions:
 - Is there any other way I could get uasyncio on my eps8266 (1mb flash) ?
 - Is there a doc that you could lead me to so that I can build firmware myself with and without modules that I can pick?

I would greatly appreciate any help

**Comment by scruss** [comment]
There isn't enough flash space to install uasyncio on a 1 Mbit ESP8266.

**Comment by robert-hh** [comment]
There is little hope that you will succeed on a 1M version of the esp8266. Better go for a 4M ESP8266 board or an ESP32. If you want to try:
The asyncio files are here: https://github.com/micropython/micropython/tree/master/extmod/asyncio
But since the heap space of the ESP8266 is small, you have to build you own firmware. To get asyncio, you have to add/change the line:
`#define MICROPY_PY_ASYNCIO              (1)`
to mpconfigboard.h, and include the line:
`include("$(MPY_DIR)/extmod/asyncio")`
into boards/manifest.py.
I tried that and it included asyncio then.
```
>>> help("modules")
__main__          btree             inisetup          random
_asyncio          builtins          io                re
_boot             collections       json              requests/__init__
_espnow           cryptolib         lwip              select
_onewire          deflate           machine           socket
_webrepl          dht               math              ssl
apa102            ds18x20           micropython       struct
array             errno             mip/__init__      sys
asyncio/__init__  esp               neopixel          time
asyncio/core      espnow            network           uasyncio
asyncio/event     flashbdev         ntptime           uctypes
asyncio/funcs     framebuf          onewire           urequests
asyncio/lock      gc                os                webrepl
asyncio/stream    hashlib           platform          webrepl_setup
binascii          heapq             port_diag         websocket
Plus any modules on the filesystem
>>> import asyncio
>>> 
```
The firmware image file is attached.

[firmware.zip](https://github.com/micropython/micropython/files/13061807/firmware.zip)

**Reply by ShocKwav3** [reply]
@robert-hh Thanks for the detailed reply. I feared as much that 1mb flash could be the issue. Well as I see from the related discussion that I can opt to exclude some modules. I appreciate that you have provided me with the firmware but can you direct me to some guide or pointers which I can follow to compile/build firmware my own? At the moment I do not know how :(

**Comment by robert-hh** [comment, marked as answer]
The instructions on how to build the firmware are at https://github.com/micropython/micropython/blob/master/ports/esp8266/README.md. Please follow them carefully. The best platform for building is Linux. WSL works as well. The actual problem I found is the esptool.py tool in the esp8266 idf, which requires Python 2. Newer versions of esptool.py work as well. Maybe using the provided Docker container avoids the Python version trouble.

**Reply by ShocKwav3** [reply]
@robert-hh Thanks a lot! I had flawless success compiling a custom firmware that includes asyncio, using the docker approach. I was also able to remove some modules such as dht, neopixel etc. At this point I have only three questions:

 - How can I exclude modules that are not listed in manifest.py?
   - For example, I removed `require('neopixel')` from it and the compiled firmware excluded the `neopixel` module
   - If I want to remove `espnow`, which is not listed there what could I do?
 - How can I use a stable version?
   - As I followed through the build instructions, it compiled using preview build.
   - What could I do to use stable builds?
 - How can I include modules that do not come with micropython
   - I can use `mpy-cross` and make a `.mpy` file and put it in the `lib` folder but I wanted to know if I could include/specify a module during the build process so it would be available by default.

**Reply by robert-hh** [reply]
Files to be included into the firmware have to be either be named in the boards/manifest.py file, or to be located in the modules subdirectory, either as a copy or a link. There you can put a link to your own modules. You do not have to precompile them. That's done in the build phase. espnow.py is in the modules directory, and there is a flag MICROPY_PY_ESPNOW in boards/ESP8266_GENERIC/mpconfigboard.mk for it, which has to be set to 0. There, you can as well disable BTREE, if you do not need it. It takes a lot of space as well. 
For including your files, you can as well add a include() statement to manifest.py pointing at a directory with your files.

You can safely use the preview versions. The so called stable version is just the preview from a certain date. But If you want to use the v1.21.0 stable, you can check it out from the repository with: `git checkout v1.21.0`. All versions and each and every interim state since the beginning are still available in the repository.
